src/main/scala/cromwell/binding/TaskOutput.scala

```diff
@@ -1,15 +1,40 @@
 package cromwell.binding
 
 import cromwell.binding.AstTools.EnhancedAstNode
+import cromwell.binding.expression.{NoFunctions, WdlStandardLibraryFunctionsType}
 import cromwell.binding.types.WdlType
-import cromwell.parser.WdlParser.Ast
+import cromwell.parser.WdlParser.{Ast, SyntaxError, Terminal}
+
+import scala.util.{Failure, Success}
 
 object TaskOutput {
-  def apply(ast: Ast, syntaxErrorFormatter: WdlSyntaxErrorFormatter): TaskOutput = {
+  def apply(ast: Ast, declarations: Seq[Declaration], syntaxErrorFormatter: WdlSyntaxErrorFormatter): TaskOutput = {
     val wdlType = ast.getAttribute("type").wdlType(syntaxErrorFormatter)
     val name = ast.getAttribute("var").sourceString
-    val expression = ast.getAttribute("expression")
-    new TaskOutput(name, wdlType, new WdlExpression(expression))
+    val expression = WdlExpression(ast.getAttribute("expression"))
+
+    // .get below because contract with the lookup() function is that it signals
+    // an error by throwing an exception which is wrapped in a Try() in the
+    // evaluator code
+    def lookup(n:String) = declarations.find(_.name == n).get.wdlType
+
+    def badCoercionException(expressionWdlType: WdlType) = throw new SyntaxError(syntaxErrorFormatter.taskOutputExpressionTypeDoesNotMatchDeclaredType(
+      ast.getAttribute("var").asInstanceOf[Terminal],
+      wdlType,
+      expressionWdlType
+    ))
+
+    expression.evaluateType(lookup, new WdlStandardLibraryFunctionsType) match {
+      case Success(expressionWdlType) if !wdlType.isCoerceableFrom(expressionWdlType) => badCoercionException(expressionWdlType)
+      case Success(expressionWdlType) =>
+        val expressionValue = expression.evaluate((s:String) => throw new Throwable("not implemented"), new NoFunctions)
+        expressionValue match {
+          case Success(value) if wdlType.coerceRawValue(value).isFailure => badCoercionException(expressionWdlType)
+          case _ =>
+        }
+      case Failure(ex) => // Log message?
```

Comment:
Per discussion with Scott this is not doing what I thought it was, it's just trying to determine the type of an expression.  That's something we should always be able to do successfully, so this really does represent an error case that I think should be logged.